File: bench/lib/bench.rb
# ruby benchmark harness -- mirrors bench/bench.g.
# bench(name) { work } auto-scales the repetition count (doubling until the run
# clears MIN_MS), then prints one line matching the gwen/python harnesses:
#     <name> ruby <reps> <ms> <checksum>
MIN_MS = 200.0

def bench(name)
  reps = 1
  loop do
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    chk = nil
    reps.times { chk = yield }
    ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000.0
    if ms >= MIN_MS
      puts format("%s ruby %d %.3f %s", name, reps, ms, chk)
      break
    end
    reps *= 2
  end
end

File: bench/lib/bench.clj
;; clojure benchmark harness -- mirrors bench/bench.l. (System/nanoTime) wall
;; clock; one line out: <name> <lang> <reps> <ms> <checksum> (println is
;; space-separated). label from BENCH_LANG (default "clojure").
(def bench-min-ms 200.0)
(def bench-lang (or (System/getenv "BENCH_LANG") "clojure"))
(defn bench [name work]
  (loop [reps 1]
    (let [t0  (System/nanoTime)
          chk (loop [i 0 c nil] (if (< i reps) (recur (inc i) (work)) c))
          ms  (/ (- (System/nanoTime) t0) 1e6)]
      (if (>= ms bench-min-ms)
        (println name bench-lang reps ms chk)
        (recur (* 2 reps))))))
